File: cpp/classic algorithms/graph/Floyd.cpp
#include <iostream>
#include <vector>
#include <climits> // INT_MAX
using namespace std;

vector<vector<int>> floyd(const vector<vector<int>>& graph) {
    int n = graph.size();
    vector<vector<int>> dist(n, vector<int>(n));

    // 复制初始图距离
    for (int i = 0; i < n; i++) {
        dist[i] = graph[i];
    }

    // Floyd-Warshall 三重循环
    for (int k = 0; k < n; k++) {
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                // dist[a][b] == INT_MAX 表示这两点之间无边链接
                if (dist[i][k] != INT_MAX && dist[k][j] != INT_MAX &&
                    dist[i][k] + dist[k][j] < dist[i][j]) {
                    dist[i][j] = dist[i][k] + dist[k][j];
                }
            }
        }
    }

    return dist;
}

int main() {
    const int INF = INT_MAX;
    vector<vector<int>> graph = {
        {0, 3, INF, 7},
        {8, 0, 2, INF},
        {5, INF, 0, 1},
        {2, INF, INF, 0}
    };

    vector<vector<int>> dist = floyd(graph);

    cout << "Floyd 最短路结果：" << endl;
    for (const auto& row : dist) {
        for (int d : row) {
            if (d == INF) cout << "INF ";
            else cout << d << " ";
        }
        cout << "\n";
    }

    return 0;
}

File: Java/ClassicAlgorithm/ShortestPath/Floyd.java
package ClassicAlgorithm.ShortestPath;

import java.util.Arrays;

public class Floyd {
    // Floyd-Warshall算法，返回所有点对最短距离
    public static int[][] floyd(int[][] graph) {
        int n = graph.length;
        int[][] dist = new int[n][n];
        for (int i = 0; i < n; i++) {
            dist[i] = Arrays.copyOf(graph[i], n);
        }
        for (int k = 0; k < n; k++) {
            for (int i = 0; i < n; i++) {
                for (int j = 0; j < n; j++) {
                    if (dist[i][k] != Integer.MAX_VALUE && dist[k][j] != Integer.MAX_VALUE
                        && dist[i][k] + dist[k][j] < dist[i][j]) {
                        dist[i][j] = dist[i][k] + dist[k][j];
                    }
                }
            }
        }
        return dist;
    }

    public static void main(String[] args) {
        int INF = Integer.MAX_VALUE;
        int[][] graph = {
            {0, 3, INF, 7},
            {8, 0, 2, INF},
            {5, INF, 0, 1},
            {2, INF, INF, 0}
        };
        int[][] dist = floyd(graph);
        System.out.println("Floyd最短路结果：");
        for (int i = 0; i < dist.length; i++) {
            System.out.println(Arrays.toString(dist[i]));
        }
    }
}
